sklearn/tree/_criterion.pyx

```diff
@@ -586,6 +586,11 @@ cdef class ClassificationCriterion(Criterion):
         elif dest[0] > upper_bound:
             dest[0] = upper_bound
 
+        if dest[0] < 0.:
+            dest[0] = 0
+        elif dest[0] > 1.:
+            dest[0] = 1.
+
         # Class proportions for binary classification must sum to 1.
         dest[1] = 1 - dest[0]
```

Comment:
OK so the bug is here. We don't expect to have a sum at 1 but a weighted count.
So we should have something like the `sum_total - dest[0]` instead.
I don't know if we have access to `sum_total` thought at this stage.

Reply:
Uhm actually this is not even `sum_total` because this is grouped by labels.

Reply:
So with my limited knowledge of the wider implementation of the constraints, I would come with the following code:

```python
        cdef float64_t total_weighted_count = dest[0] + dest[1]

        if dest[0] < lower_bound:
            dest[0] = lower_bound
        elif dest[0] > upper_bound:
            dest[0] = upper_bound

        dest[1] = total_weighted_count - dest[0]
```

Reply:
Thanks a lot for looking into this @glemaitre !
> We don't expect to have a sum at 1 but a weighted count

I was not aware of this obviously, so this is a big mistake in the monotonic constraints implementation indeed !

> Uhm actually this is not even sum_total because this is grouped by labels.

Monotonic constraints are only supported with single label, so this should not be an issue.

Note that `lower_bound` and `upper_bound` are positive class frequencies so we would need to modify your solution to 
``` python
        cdef float64_t total_weighted_count = dest[0] + dest[1]

        if dest[0] < lower_bound * total_weighted_count:
            dest[0] = lower_bound * total_weighted_count
        elif dest[0] > upper_bound * total_weighted_count:
            dest[0] = upper_bound * total_weighted_count

        dest[1] = total_weighted_count - dest[0]
```

Reply:
The patch almost works !

`test_monotonic_constraints_classifications` currently fails due to floating point arithmetic fluctuations. The tests pass if the predicted probabilities are cast to `np.float32` so the overall logic is valid, which is good news.

I'm looking into the `float64` precision matter.

Reply:
Nice. Now we only have to deal with the 25 CIs with some 32-bits arch ;)

If we have some casting in float32, we can probably play with the relative tolerance.

Reply:
My understanding is that the numerical noise comes from the predictions on clipped leaves:
```python
(lower_bound * total_weighted_count) / total_weighted_count
```
which **do vary with `total_weighted_count`** (in floating point arithmetic) close to `lower_bound`. So two leaves can have values (before clipping) trespassing the same constraints, say `lower_bound`, but with different `total_weighted_count` and their predictions fluctuate a bit and sometimes break the monotonicity.

I found a solution that works by loosing precision on `lower_bound` (and `upper_bound`) before performing the scaling:
```python
        cdef float64_t total_weighted_count = dest[0] + dest[1]
        cdef float64_t scaled_lower_bound = (<float32_t>lower_bound) * total_weighted_count
        cdef float64_t scaled_upper_bound = (<float32_t>upper_bound) * total_weighted_count

        if dest[0] < scaled_lower_bound:
            dest[0] = scaled_lower_bound
        elif dest[0] > scaled_upper_bound:
            dest[0] = scaled_upper_bound

        dest[1] = total_weighted_count - dest[0]

```
However, it does have limitations:
- `sample_weights` that have a precision of `float32` or more (tested `float16` and this is OK)
- `n_samples` too close to `2**32`, so larger than a few hundred of millions (not for `n_samples` tested beyond a few millions though...)